protocols/relay/src/transport.rs

```diff
@@ -0,0 +1,432 @@
+    #[pin]
+    inner_listener: Option<<T as Transport>::Listener>,
+    from_behaviour: Arc<Mutex<mpsc::Receiver<BehaviourToTransportMsg>>>,
+
+    msg_to_behaviour: Option<BoxFuture<'static, Result<(), mpsc::SendError>>>,
+    report_listen_addr: Option<Multiaddr>,
+}
+
+impl<T: Transport> Stream for RelayListener<T> {
+    type Item = Result<
+        ListenerEvent<RelayedListenerUpgrade<T>, EitherError<<T as Transport>::Error, RelayError>>,
+        EitherError<<T as Transport>::Error, RelayError>,
+    >;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let this = self.project();
+
+        if let Some(msg) = this.msg_to_behaviour {
+            match Future::poll(msg.as_mut(), cx) {
+                Poll::Ready(Ok(())) => *this.msg_to_behaviour = None,
+                Poll::Ready(Err(e)) => return Poll::Ready(Some(Err(EitherError::B(e.into())))),
+                Poll::Pending => {}
+            }
+        }
+
+        if let Some(listener) = this.inner_listener.as_pin_mut() {
+            match listener.poll_next(cx) {
+                Poll::Ready(Some(Err(e))) => return Poll::Ready(Some(Err(EitherError::A(e)))),
+                Poll::Ready(Some(Ok(ListenerEvent::Upgrade {
+                    upgrade,
+                    local_addr,
+                    remote_addr,
+                }))) => {
+                    return Poll::Ready(Some(Ok(ListenerEvent::Upgrade {
+                        upgrade: RelayedListenerUpgrade::Inner(upgrade),
+                        local_addr,
+                        remote_addr,
+                    })))
+                }
+                Poll::Ready(Some(Ok(ListenerEvent::NewAddress(addr)))) => {
+                    return Poll::Ready(Some(Ok(ListenerEvent::NewAddress(addr))))
+                }
+                Poll::Ready(Some(Ok(ListenerEvent::AddressExpired(addr)))) => {
+                    return Poll::Ready(Some(Ok(ListenerEvent::AddressExpired(addr))))
+                }
+                Poll::Ready(Some(Ok(ListenerEvent::Error(err)))) => {
+                    return Poll::Ready(Some(Ok(ListenerEvent::Error(EitherError::A(err)))))
+                }
+                Poll::Ready(None) => return Poll::Ready(None),
+                Poll::Pending => {}
+            }
+        }
+
+        match this.from_behaviour.lock().unwrap().poll_next_unpin(cx) {
+            Poll::Ready(Some(BehaviourToTransportMsg::IncomingRelayedConnection {
+                stream,
+                src_peer_id,
+                relay_addr,
+            })) => {
+                return Poll::Ready(Some(Ok(ListenerEvent::Upgrade {
+                    upgrade: RelayedListenerUpgrade::Relayed(Some(stream)),
+                    local_addr: relay_addr.with(Protocol::P2pCircuit),
+                    remote_addr: Protocol::P2p(src_peer_id.into()).into(),
+                })));
+            }
+            Poll::Ready(None) => unimplemented!(),
```

Comment:
Do I understand correctly that this can only happen when the behaviour is dropped? Since there are similar panics in the behaviour when the transport gets dropped, I wonder if this can not cause trouble, as one of them is always destined to be dropped first?

Reply:
Absolutely correct. This is problematic, especially with the reworked listener logic (https://github.com/libp2p/rust-libp2p/pull/1838#discussion_r582262518). I will change the code to terminate the listener instead. Thanks for highlighting this.